### ISSUE TITLE
v0.9 Preview

### ISSUE THREAD

**Issue by felixhayashi** [felixhayashi/TW5-TiddlyMap#161]
Hi @ all,

I uploaded a demo to http://tiddlymap.org/experimental/index.html.

**This demo is experimental** (as the url already suggests)! **Do not install this yet! I'll release v0.9 officially in this or the next week after thorough testing.** Also I need to wait till https://github.com/almende/vis/issues/1226 is fixed at the vis.js upstream.

**v0.9 brings many changes for styling and tweaking nodes and the graph.** See version info tiddler of the demo wiki for more info.

I suggest you have a look at this first to understand the various levels of styling: http://tiddlymap.org#Styling%20Nodes

I also released a (rather poorly done) youtube video that is worth watching (see video description to skip to the good part): https://youtu.be/DC7r744NCCg . The video shows how to style groups of nodes. The experimental wiki also contains the example, so you can play with it.

I appreciate any feedback :)

-Felix

**Comment by felixhayashi** [comment]
![2015-09-02 00 14 58](https://cloud.githubusercontent.com/assets/4307137/9618540/b2914f00-5107-11e5-829e-33751e81ef6a.png)

**Comment by Peter-Miller** [comment]
When applying global style from the modal dialog, the help text for the Quit button would be more accurately described as "Cancel unsaved changes and exit".

**Comment by Peter-Miller** [comment]
Selecting a node highlights it using the previous (or global?) style. Deselecting it reverts to the correct style.

**Comment by Peter-Miller** [comment]
Using the node-type manager (General tab) and the All Tiddlers view, I define a scope for my new style (a new background colour) based on a tag I just defined. I save, the map updates and the node changes colour. I'm assuming this is a global type. 

I then change the view in the manager (General tab) and Save but the node colour doesn't revert in the original All Tiddlers view despite this now being a local type.

**Comment by Peter-Miller** [comment]
Is there a way to defer rendering when working with large maps? It can be a little time-consuming.

**Comment by Peter-Miller** [comment]
It might be worth showing an alert when the Create button in the node-manager is pressed without a name being defined -- when caffeine-deficient it's quite easy to search for a style and, when it's not found, click Create and assume you're using the search term as the name for the new style type.

**Comment by Peter-Miller** [comment]
Now getting following error when attempting to access demo wiki despite restarting browser (Firefox)

Error: Permission denied to access property "type"

No problem with same wiki in Chrome.

**Comment by felixhayashi** [comment]
Much thanks for this detailed feedback @Peter-Miller 

> When applying global style from the modal dialog, the help text for the Quit button would be more accurately described as "Cancel unsaved changes and exit".

**Comment by felixhayashi** [comment]
I'll go over your points later today!

**Comment by felixhayashi** [comment]
Ok here we go :)

> When applying global style from the modal dialog, the help text for the Quit button would be more accurately described as "Cancel unsaved changes and exit".

This would really count for all modals in all situations since tiddlymap modals only save changes when ok or save is pressed. I therefore decided to add a tooltip to all buttons. If I add "Cancel unsaved changes and exit" only to a button in a single dialog, then people will think, in dialogs that it is not the case.

> Selecting a node highlights it using the previous (or global?) style. Deselecting it reverts to the correct style.

At the moment, you can define how a node should be highlighted on many levels. When no node specic style is defined, TiddlyMap looks at the group styles (added via node-types). If no node-types are defined, the global default style ("configure tiddlymap") or local default style ("configure view") is applied.

**Suggestion:** I think I will completely remove the vis configuration option to specify the highlighting style then… The highlighted color will be the same as the backgroundcolor of the node. This is also how it used to be in pre v0.9…

> Using the node-type manager (General tab) and the All Tiddlers view, I define a scope for my new style (a new background colour) based on a tag I just defined. I save, the map updates and the node changes colour. I'm assuming this is a global type.
> I then change the view in the manager (General tab) and Save but the node colour doesn't revert in the original (All Tiddlers) view despite this now being a local type for a different view.

Yes, this is the bug. After the dialog closes, I need to clear the node styles of the view to remove the previously loaded styles. Same for when chaning a view. But if you change the sidebar tab and move back to the "map" tab you will see, that the node style really only applies to view it is bound too. 

This bug is one primary reason, why I did not publish v0.9 yet. However, the visjs bug mentioned in the description (https://github.com/almende/vis/issues/1226) prevents me from doing so.

> Is there a way to defer rendering when working with large maps? It can be a little time-consuming.

I did not understand this one. You mean coloring or adding styles?

I still need to improve TiddlyMap speed. Over the past months, I added a lot of features and I think I will dedicate the next time to work on speed.

> It might be worth showing an alert when the Create button in the node-manager is pressed without a name being defined -- when caffeine-deficient it's quite easy to search for a style and, when it's not found, click Create and assume you're using the search term as the name for the new style type.

This also happens to me so often!!!! And I thought it was only me :) 

**Suggestion:** I will add this as a feature. When no style is found in the search, I will show a button in the search result list next to the text "Sorry, no results" which has the label "Create a new node-type with this name".

> > Error: Permission denied to access property "type"
> Update: The following doesn't appear to be reproducible. Please ignore.

Please never hesitate to report bugs even if they are not reproducible. In this case, I know this bug. It is a firefox issue that (in my case) sometimes occurs when you freshly open a browser and detach window tabs. It started happening with the latest firefox versions and I honestly don't know how to solve it. I guess I'll simply swallow the error message.

Thanks again for your feedback! I'll update the experimental online version soon to reflect some of the changes you suggested.

-Felix

**Comment by Peter-Miller** [comment]
Re rendering, I was referring to the delay that occurs when you save and quit from the node-manager. Your comment re speed is a perfect response.

**Comment by Peter-Miller** [comment]
Occasionally seeing this in Firefox.
![tmap1](https://cloud.githubusercontent.com/assets/10705585/9634179/651ff830-5188-11e5-87fe-b914c0f8bc6b.png)

**Comment by Peter-Miller** [comment]
Got this in Firefox for node background. Can't recall exact sequence but I think it was "load last" (no last available), fiddle with settings, Save, Quit and immediate re-edit of node background.
![tmap2](https://cloud.githubusercontent.com/assets/10705585/9634226/959aa384-5188-11e5-97ab-5ddec78708f2.png)

**Comment by felixhayashi** [comment]
Thanks @Peter-Miller 

I'm still working on some modifications and will release another preview version soon!

-Felix

**Comment by felixhayashi** [comment]
Hi @Peter-Miller 

updated the preview version. There are still some remaining bugs (e.g. the arrows are sometimes not displayed correctly) but at least an important one is fixed now (see below). I also did some improvements.

> I then change the view in the manager (General tab) and Save but the node colour doesn't revert in the original (All Tiddlers) view despite this now being a local type for a different view. 

This is fixed now.

> When no style is found in the search, I will show a button in the search result list next to the text "Sorry, no results" which has the label "Create a new node-type with this name".

This has been added.

> Re rendering, I was referring to the delay that occurs when you save and quit from the node-manager. Your comment re speed is a perfect response.

I hope I can improve TiddlyMap performance once all the functionality has settled a bit. But generally speaking, node-types increase calculation time because I need to recalculate their scope on every tiddler change since the change could mean that a tiddler is now out of scope or newly inside a scope. But I am sure I come up with a good idea.

By the way, working with many small maps speeds up the process in comparison to working large maps. I hope I find a good solution for #102 soon so modular map design becomes more attractive.

---

Apart from that I added support for Font Awesome and the demo contains some examples (e.g. the view "Overview" contains some FA icons…).

-Felix
